include/adc.h

```diff
@@ -27,119 +27,225 @@ extern "C" {
  */
 
 /**
- * @brief ADC driver Sequence entry
- *
- * This structure defines a sequence entry used
- * to define a sample from a specific channel.
+ * @brief ADC sampling definition structure.
  */
-struct adc_seq_entry {
-	/** Clock ticks delay before sampling the ADC. */
-	s32_t sampling_delay;
+struct adc_sampling {
+	/**
+	 * Pointer to a buffer where the samples are to be written.
+	 * It must be of proper size, taking into account the number of samples
+	 * and used ADC resolution.
+	 */
+	void *buffer;
 
-	/** Buffer pointer where the sample is written.*/
-	u8_t *buffer;
+	/**
+	 * Channels to be used for getting consecutive samples.
+	 */
+	const u8_t *channel_ids;
 
-	/** Length of the sampling buffer.*/
-	u32_t buffer_length;
+	/**
+	 * Number of samples to read in the sampling.
+	 */
+	u8_t sample_count;
+};
 
-	/** Channel ID that should be sampled from the ADC */
-	u8_t channel_id;
+enum adc_trigger
+{
+	ADC_TRIGGER_DELAY_TICKS,
+	ADC_TRIGGER_DELAY_US,
+	ADC_TRIGGER_DELAY_MS,
+
+	ADC_TRIGGER_SAMPLE_RATE,
+};
+
+/* Forward declaration of the adc_sequence structure. */
+struct adc_sequence;
+
+typedef int (*adc_sequence_callback_t)(struct device *dev,
+				       const struct adc_sequence *sequence,
+				       u8_t sampling_number);
+
+struct adc_sequence_continuation {
+	/** Number of consecutive samplings to perform. */
+	u8_t sampling_count;
+
+	/** Type of trigger that starts each sampling. */
+	enum adc_trigger trigger_type;
+	/** Parameter for the selected trigger, number of ticks, etc. */
+	u32_t trigger_param;
 
-	u8_t stride[3];
+	/**
+	 * Callback to be called after each sampling is done.
+	 * Optional - set to NULL if it is not needed.
+	 */
+	adc_sequence_callback_t callback;
 };
 
 /**
- * @brief ADC driver Sequence table
- *
- * This structure defines a list of sequence entries
- * used to execute a sequence of samplings.
+ * @brief ADC sampling definition structure.
+ */
+struct adc_sequence {
+	/**
+	 * ADC resolution.
+	 * For single-ended channels the sample values are from range:
+	 *   0 .. 2^resolution - 1,
+	 * for differential ones:
+	 *  -2^(resolution-1) .. 2^(resolution-1) - 1.
+	 */
+	u8_t resolution;
+
+	/**
+	 * Oversampling setting.
+	 * Each sample is averaged from 2^oversampling results.
+	 */
+	u8_t oversampling;
+
```

Comment:
I would reorder the attributes so these 2 u8_t would be at the end of the struct.